src/Compilers/Core/Portable/Syntax/SyntaxTree.cs

```diff
@@ -343,6 +352,16 @@ internal Cci.DebugSourceInfo GetDebugSourceInfo()
         /// </summary>
         public abstract SyntaxTree WithFilePath(string path);
 
+        /// <summary>
+        /// Returns a new tree whose <see cref="DiagnosticOptions" /> are the specifed value and other properties are copied
+        /// from the current tree.
+        /// </summary>
+        /// <param name="options">
+        /// A mapping from diagnostic id to diagnostic reporting level. The diagnostic ID may be case-sensitive depending
+        /// on the language.
+        /// </param>
+        public abstract SyntaxTree WithDiagnosticOptions(IEnumerable<(string, ReportDiagnostic)> options);
```

Comment:
>(string, ReportDiagnostic) [](start = 69, length = 26)

I've heard some concerns in the past about using tuples in public API.

Reply:
I can change it to KVP, but I'm curious what those issues would be. @jcouv Do you remember any issues here?

Reply:
Personally, i would use KVP.  Not because of any problems with tuples, but because it's how we've already done things in ParseOptions.  i.e. ```WithFeatures(IEnumerable<KeyValuePair<string, string>> features)```.

Also, by accepting KVP you naturally accept dictionaries, meaning you don't need all the code to convert to/from ```IE<(...)>``` to dictionary and back again.  

If we didn't have existing KVP apis here, i would likely feel differently.  But better to have this all workign the same way IMO.